File: app/src/test/java/alon/soffer/rachel_sandwiches/AppTest.java
package alon.soffer.rachel_sandwiches;

import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;


import com.google.android.material.floatingactionbutton.FloatingActionButton;

import junit.framework.TestCase;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.android.controller.ActivityController;
import org.robolectric.annotation.Config;

import java.util.ArrayList;

import static org.mockito.ArgumentMatchers.eq;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = 28)
public class MainActivityTest extends TestCase {

    private ActivityController<NewOrderActivity> newOrderActivityActivityController;
    private ActivityController<WaitingActivity> waitingActivityActivityController;
    private ActivityController<InProgressActivity> inProgressActivityActivityController;
    private ActivityController<ReadyActivity> readyActivityActivityController;


    @Before
    public void setup(){
        newOrderActivityActivityController = Robolectric.buildActivity(NewOrderActivity.class);
        waitingActivityActivityController = Robolectric.buildActivity(WaitingActivity.class);
        inProgressActivityActivityController = Robolectric.buildActivity(InProgressActivity.class);
        readyActivityActivityController = Robolectric.buildActivity(ReadyActivity.class);
    }

    @Test
    public void when_orderButtonClicked_then_waitingActivityIsLunched() {
        // setup
        newOrderActivityActivityController.create().visible();
        NewOrderActivity activityUnderTest = newOrderActivityActivityController.get();

        View orderButton = activityUnderTest.findViewById(R.id.makeOrderButton);
        orderButton.performClick();

        // verify
        WaitingActivity waitingActivity =
        assertTrue(userInput.isEmpty());
    }

}

File: app/src/test/java/alon/soffer/rachel_sandwiches/AppTest.kt
package alon.soffer.rachel_sandwiches

import android.content.ComponentName
import android.os.Looper
import junit.framework.TestCase
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows
import org.robolectric.Shadows.shadowOf
import org.robolectric.android.controller.ActivityController
import org.robolectric.annotation.Config
import org.robolectric.annotation.LooperMode

@RunWith(RobolectricTestRunner::class)
@LooperMode(LooperMode.Mode.PAUSED)
@Config(sdk = [28])
class AppTest : TestCase() {

    @Test
    fun when_newOrderUploadsToDB_then_waitingActivityIsLunched() {
        // setup
        val newOrderActivityActivityController = Robolectric.buildActivity(NewOrderActivity::class.java)
        newOrderActivityActivityController.create().visible()
        val activityUnderTest = newOrderActivityActivityController.get()
        val shadow = shadowOf(activityUnderTest)
        activityUnderTest.onDbSuccess("")

        // verify
        val component = shadow.nextStartedActivity.component
        assertEquals(ComponentName(activityUnderTest, WaitingActivity::class.java), component)
        assertTrue(activityUnderTest.isFinishing)
    }


    @Test
    fun when_orderMarkedDone_then_goToNewOrderAndNoCurrentOrderInApp() {
        // setup
        val readyActivityController = Robolectric.buildActivity(ReadyActivity::class.java)
        readyActivityController.create().visible()
        val activityUnderTest = readyActivityController.get()
        val shadow = shadowOf(activityUnderTest)
        val order = SandwichOrder()
        order.status = RachelApplication.READY
        activityUnderTest.onStatusChangedToDone(order)

        // verify
        val component = shadow.nextStartedActivity.component
        assertEquals(ComponentName(activityUnderTest, NewOrderActivity::class.java), component)
        assertTrue(activityUnderTest.isFinishing)
    }

    @Test
    fun when_appOpensWithOrderInProgress_then_InProgressActivityOpens(){
        val welcomeActivityController = Robolectric.buildActivity(WelcomeActivity::class.java)
        welcomeActivityController.create().visible()
        val activityUnderTest = welcomeActivityController.get()

        val shadow = shadowOf(activityUnderTest)
        val order = SandwichOrder()
        order.status = RachelApplication.IN_PROGRESS
        activityUnderTest.onGotOrderFromDb(order)

        // verify
        val component = shadow.nextStartedActivity.component
        assertEquals(ComponentName(activityUnderTest, InProgressActivity::class.java), component)
        assertTrue(activityUnderTest.isFinishing)
    }
}